develop-guide/writeparser/writeparser.rst

```diff
@@ -1,47 +1,558 @@
+
+If you intend to contribute your parser to the pyATS project, then you must
+build and implement a schema class as it results in the following benefits:
+
+* *Time-saving*: You can quickly see the data structure without having to read hundreds of lines of regex output. This saves you time when troubleshooting.
+* *Future-proof* and *robust*: When you or others modify code, you're less likely to break something.
+* *Scalable*: It's more efficient to modify a schema than to have multiple developers working with just the regex output.
+
+
+.. _creating_a_schema:
+
+5.1 Creating a schema
+=====================
+
+This section covers the making of your own new schema.
+
+To start creating a new schema, you'll first need to gather device output.
+Once you have some, you can identify the keys that you'll need in a number of ways, including:
+
+| :ref:`5.1.1<keys_from_show_command>` Examining the ``show`` command output directly.
+| :ref:`5.1.2<keys_from_XML>` Using the XML output option with a ``show`` command to display the output as key-value pairs.
+| :ref:`5.1.3<keys_from_YANG>` Using the YANG data model to identify the relevant keys.
+
+.. note:: You can also :ref:`create a schema based on an existing model <schemabasedonanexistingmodel>`
+
+.. tip:: You can find more examples of schemas :ref:`here <schemaexamples>`
+
+
+.. _keys_from_show_command:
+
+5.1.1 Identifying keys directy from show command output
+-------------------------------------------------------
+
+Writing a schema based on device output can range from straightforward to
+surprisingly complex. Having multiple examples of device output for a given
+command can be incredibly helpful when designing a schema.
+
+For example, the output from the `show track` command for IOSXE can look like
+this:
+
+.. code-block:: python
+
+    Track 1
+    Interface GigabitEthernet3.420 line-protocol
+    Line protocol is Up
+        1 change, last change 00:00:27
+    Tracked by:
+        VRRP GigabitEthernet3.420 10
+
+Or it can look like this:
+
+.. code-block:: python
+
+    Track 2
+    IP route 10.21.12.0 255.255.255.0 reachability
+    Reachability is Down (no ip route), delayed Up (1 sec remaining) (connected)
+        1 change, last change 00:00:24
+    Delay up 20 secs, down 10 secs
+    First-hop interface is unknown (was Ethernet1/0)
+    Tracked by:
+        HSRP Ethernet0/0 3
+        HSRP Ethernet0/1 3
+
+Or even this:
+
+.. code-block:: python
+
+    Track 1
+    IP route 172.16.52.0 255.255.255.0 metric threshold
+    Metric threshold is Down (no route)
+        1 change, last change 00:00:35
+    Metric threshold down 255 up 254
+    Delay up 2 secs, down 1 sec
+    First-hop interface is unknown
+
+Comparing these three outputs reveals a lot about the structure of the information
+given by that particular show command. Using these output examples, a schema that
+looks like this can be created:
+
+.. code-block:: python
+
+    class ShowTrackSchema(MetaParser):

```

Comment:
Is this a real schema,  can you make sure with Takashi it is a good one?

Reply:
Yes, real schema. Made it myself recently. Takashi helped me to write it.

I used it because I didn't want a super basic example ( like tabular output) but I also didn't want anything too overwhelming for someone new to Python.

I'll check with Takashi anyway

Reply:
Asked Takashi, too busy today. Will bring this up with him next week